telefunc/node/server/shield/codegen/generateShield.ts

```diff
@@ -0,0 +1,379 @@
+export { generateShield }
+export { replaceShieldTypeAlias }
+export { printGenStatus }
+export { testGenerateShield }
+
+import { Project, VariableDeclarationKind, SourceFile, getCompilerOptionsFromTsConfig } from 'ts-morph'
+import {
+  assert,
+  assertUsage,
+  assertWarning,
+  assertModuleScope,
+  objectAssign,
+  unique,
+  projectInfo
+} from '../../../utils'
+import fs from 'fs'
+import path from 'path'
+import pc from 'picocolors'
+
+type GeneratedShield = {
+  telefuncFilePath: string
+  telefunctionName: string
+  failed: boolean
+  project: Project & { tsConfigFilePath: null | string }
+}
+
+assertModuleScope('codegen/generateShield.ts')
+const generatedShields: GeneratedShield[] = []
+const projects: Record<string, Project> = {}
+
+function generateShield(telefuncFileCode: string, telefuncFilePath: string): string {
+  const { project, telefuncFileSource, shieldGenSource } = getProject(telefuncFilePath, telefuncFileCode)
+  // We should preserve prior `telefuncFileCode` transformations
+  telefuncFileSource.replaceWithText(telefuncFileCode)
+  return generate({ project, telefuncFileSource, shieldGenSource, telefuncFilePath })
+}
+
+function getProject(telefuncFilePath: string, telefuncFileCode: string) {
+  const tsConfigFilePath = findTsConfig(telefuncFilePath)
+  const key = tsConfigFilePath ?? '__no_tsconfig'
+  const typeToShieldFilePath = path.join(getFilsystemRoot(), '__telefunc_typeToShield.ts')
+
+  if (!projects[key]) {
+    let project: Project
+    if (!tsConfigFilePath) {
+      project = projects[key] = new Project({
+        compilerOptions: {
+          strict: true
+        }
+      })
+
+      assert(project.getSourceFiles(telefuncFilePath).length === 0)
+      project.createSourceFile(
+        telefuncFilePath,
+        telefuncFileCode,
+        // Why do we need `overwrite`? Is that a `ts-morph` bug?
```

Comment:
Because `project.save()` would overwrite any changes to the file that exists in the filesystem at that path. It is not a bug but a better API would have been `project.save({overwrite: true})`.

Reply:
I wouldn't have expected `overwrite` to be needed because as the assert above `assert(project.getSourceFiles(telefuncFilePath).length === 0)` states, the project doesn't have that source file, so it doesn't seem to make sense to need to overwrite something that doesn't exist in the first place?

Reply:
The file exists in the filesystem though.

Reply:
I see. I removed the comment.